File: Nursery/UI/Theme.swift
import SwiftUI
import UIKit

extension UIColor {
    convenience init(hex: UInt32) {
        self.init(red: CGFloat(hex >> 16 & 0xFF) / 255, green: CGFloat(hex >> 8 & 0xFF) / 255,
                  blue: CGFloat(hex & 0xFF) / 255, alpha: 1)
    }
}

extension Color {
    /// A colour with one value for light mode and one for dark mode.
    init(light: UInt32, dark: UInt32) {
        self.init(uiColor: UIColor { $0.userInterfaceStyle == .dark ? UIColor(hex: dark) : UIColor(hex: light) })
    }
}

/// Two palettes. Light ("morning nursery"): a soft dawn blue-grey with a warm amber accent.
/// Dark ("night"): deep navy with warm moonlight; the screen is often the only light in a bedroom.
/// The picture area and Night mode are always dark. Red means a fault, never a normal state.
/// The state of the room has its own colours (teal, amber, wine, graphite): see `field(for:dim:)`.
enum Theme {
    static let skyTop = Color(light: 0xF6F8FC, dark: 0x090E1F)
    static let skyBottom = Color(light: 0xE9EDF6, dark: 0x121831)
    static let card = Color.primary.opacity(0.05)
    static let cardStroke = Color.primary.opacity(0.08)
    /// The fill of an active control. It stays yellow in both modes, with dark text on it.
    static let moon = Color(light: 0xF5CF6E, dark: 0xF7D98C)
    /// The tint of text, icons, and system controls. Deeper in light mode, for contrast.
    static let accent = Color(light: 0xA86B0C, dark: 0xF7D98C)
    static let calm = Color(light: 0x1C9A6C, dark: 0x73E3B8)          // Live, listening.
    static let warn = Color(light: 0xD27D12, dark: 0xFFB861)          // Connecting, a sound now.
    static let alarm = Color(light: 0xD8433D, dark: 0xFF6B6B)         // The sound is lost.
    static let loud = Color(light: 0xE0613F, dark: 0xFF9973)
    static let middle = Color(light: 0xE3A020, dark: 0xF7D98C)
    static let glowNeutral = Color(light: 0x9AA3B8, dark: 0x5A6070)
    static let secondaryText = Color.secondary

    static var background: some View {
        LinearGradient(colors: [skyTop, skyBottom], startPoint: .top, endPoint: .bottom).ignoresSafeArea()
    }

    /// The colour of one bar of the waveform: teal, amber, wine, the same hues as the state fields.
    /// So the waveform and the field agree.
    static func level(_ v: Float) -> Color {
        switch v {
        case ..<0.35: return stateCalm
        case ..<0.7: return stateSound
        default: return stateCry
        }
    }

    // MARK: The state of the room

    // One colour per state. The field is the whole background of the glance view, the band in the
    // picture view, the frame of the video. Red (`alarm`) stays for a fault: the cry is wine.
    // Light: the bright set, for a lit room. Dark appearance and auto-dim: a darker set with white
    // type on all fields, so a sigh at night never doubles the light in the room.

    /// The state hues on a page without a field (waveform bars, the hour strip, marks).
    static let stateCalm = Color(light: 0x00A1A0, dark: 0x2BB5B3)
    static let stateSound = Color(light: 0xE8A92A, dark: 0xFBC040)
    static let stateCry = Color(light: 0xA81233, dark: 0xC64B70)
    /// Dark type on the amber field. White there is about 2:1, ink is 10:1.
    static let ink = Color(uiColor: UIColor(hex: 0x1B1B1F))
    /// The ribbon under the word: muted, volume low, a short gap. Never a field colour.
    static let ribbon = Color(uiColor: UIColor(hex: 0x5C6B8A))
    /// The wifi glyph of "Nehlídá": the dark-mode alarm red, 4:1 on graphite.
    static let lostGlyph = Color(uiColor: UIColor(hex: 0xFF6B6B))

    private static func fieldHex(_ state: RoomState, dim: Bool) -> UInt32 {
        switch state {
        case .calm: dim ? 0x117376 : 0x00A1A0
        case .sound: dim ? 0x78662E : 0xFBC040
        case .cry: dim ? 0x570F29 : 0xA81233
        case .lost, .connecting: dim ? 0x2A2C33 : 0x3A3D45
        }
    }

    /// The field colour of a state. `dim` is true in the dark appearance and after auto-dim.
    /// Not a dynamic colour on purpose: the views on a field override the colour scheme
    /// (white type = dark scheme), and the field must not follow that.
    static func field(for state: RoomState, dim: Bool) -> Color {
        Color(uiColor: UIColor(hex: fieldHex(state, dim: dim)))
    }

    /// The core of the field: 8 % lighter. The field is one static radial gradient from it.
    static func fieldCore(for state: RoomState, dim: Bool) -> Color {
        let hex = fieldHex(state, dim: dim)
        func up(_ c: UInt32) -> CGFloat { let v = CGFloat(c & 0xFF) / 255; return v + (1 - v) * 0.08 }
        return Color(uiColor: UIColor(red: up(hex >> 16), green: up(hex >> 8), blue: up(hex), alpha: 1))
    }

    /// True when the type on the field is dark (only the bright amber). The status bar and the
    /// colour scheme of the controls on the field follow it.
    static func fieldIsLight(_ state: RoomState, dim: Bool) -> Bool { state == .sound && !dim }

    /// The glyph, the word and the labels on a field: white, or ink on the bright amber.
    static func onField(for state: RoomState, dim: Bool = false) -> Color {
        fieldIsLight(state, dim: dim) ? ink : .white
    }

    /// Small text on a field: white 90 %, ink 70 % on amber. About 4.5:1.
    static func onFieldSecondary(for state: RoomState, dim: Bool = false) -> Color {
        fieldIsLight(state, dim: dim) ? ink.opacity(0.7) : .white.opacity(0.9)
    }

    /// The state colour on black, for Night mode. The cry is a lighter wine there, else it is lost in black.
    static func nightAccent(for state: RoomState) -> Color {
        switch state {
        case .calm: Color(uiColor: UIColor(hex: 0x00A1A0))
        case .sound: Color(uiColor: UIColor(hex: 0xFBC040))
        case .cry: Color(uiColor: UIColor(hex: 0xC64B70))
        case .lost: lostGlyph
        case .connecting: .white
        }
    }

    /// The glyph of a state. The silhouette around it (ring, bare bars, disc) is in `StateGlyph`.
    static func symbol(for state: RoomState) -> String {
        switch state {
        case .connecting: "antenna.radiowaves.left.and.right"
        case .calm: "moon.zzz.fill"
        case .sound, .cry: "waveform"
        case .lost: "wifi.exclamationmark"
        }
    }

    static func color(for status: NurseryActivityAttributes.Status) -> Color {
        switch status {
        case .listening: calm
        case .silent: accent
        case .connecting: warn
        case .lost: alarm
        }
    }

    static func symbol(for status: NurseryActivityAttributes.Status) -> String {
        switch status {
        case .listening: "ear.fill"
        case .silent: "bell.badge.fill"
        case .connecting: "antenna.radiowaves.left.and.right"
        case .lost: "exclamationmark.triangle.fill"
        }
    }
}

/// The loudness of the room in words. The engine holds a word for a moment before it changes,
/// so the large headline does not flicker with each syllable.
enum RoomLevel: Int, Comparable {
    case quiet, some, loud, veryLoud

    init(_ v: Float) {
        switch v {
        case ..<0.15: self = .quiet
        case ..<0.45: self = .some
        case ..<0.75: self = .loud
        default: self = .veryLoud
        }
    }

    /// A typical level for the word, for a glow that follows the words, not each 0.1 s tick.
    var value: Float {
        switch self {
        case .quiet: 0.05
        case .some: 0.3
        case .loud: 0.6
        case .veryLoud: 0.9
        }
    }

    var title: String {
        switch self {
        case .quiet: "Ticho"
        case .some: "Slabé zvuky"
        case .loud: "Hlasitý zvuk"
        case .veryLoud: "Velmi hlasitý zvuk"
        }
    }

    static func < (a: RoomLevel, b: RoomLevel) -> Bool { a.rawValue < b.rawValue }
}

/// A dot that stays at full colour, with a soft ring that grows and fades.
/// The loop runs in its own phase animator, so it never animates the views around it.
struct PulseDot: View {
    let color: Color
    var size: CGFloat = 8
    var animated = true
    @Environment(\.accessibilityReduceMotion) private var reduceMotion
    /// It pulses for some seconds after it appears or changes, then it stays still. A pulse that
    /// never stops keeps SwiftUI drawing 60 frames a second for the whole night.
    @State private var pulsing = true

    var body: some View {
        Circle()
            .fill(color)
            .frame(width: size, height: size)
            .background {
                if animated && !reduceMotion && pulsing {
                    Circle()
                        .stroke(color, lineWidth: 1.5)
                        .phaseAnimator([false, true]) { ring, grown in
                            ring.scaleEffect(grown ? 2.4 : 1).opacity(grown ? 0 : 0.7)
                        } animation: { grown in
                            grown ? .easeOut(duration: 1.4) : nil
                        }
                }
            }
            .accessibilityHidden(true)
            .task(id: animated) {
                pulsing = true
                try? await Task.sleep(for: .seconds(8))
                pulsing = false
            }
    }
}

extension Font {
    static func rounded(_ style: Font.TextStyle, _ weight: Font.Weight = .regular) -> Font {
        .system(style, design: .rounded).weight(weight)
    }
}

/// A card with a soft glass edge.
struct Card<Content: View>: View {
    @ViewBuilder var content: Content
    var body: some View {
        content
            .padding(16)
            .frame(maxWidth: .infinity, alignment: .leading)
            .background(Theme.card, in: RoundedRectangle(cornerRadius: 24, style: .continuous))
            .overlay(RoundedRectangle(cornerRadius: 24, style: .continuous).strokeBorder(Theme.cardStroke))
    }
}

/// A round glass button for the controls on top of the picture and in the header.
struct GlassCircleButton: View {
    let symbol: String
    var size: CGFloat = 40
    var tint: Color = .primary
    let label: String
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            Image(systemName: symbol)
                .font(.system(size: size * 0.42, weight: .semibold))
                .foregroundStyle(tint)
                .frame(width: size, height: size)
                .glass(in: Circle(), interactive: true)
        }
        .buttonStyle(PressScale())
        .accessibilityLabel(label)
    }
}

/// A button that shrinks a little when pressed. The eye sees the press at once.
struct PressScale: ButtonStyle {
    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .scaleEffect(configuration.isPressed ? 0.92 : 1)
            .animation(.spring(response: 0.25, dampingFraction: 0.7), value: configuration.isPressed)
    }
}

enum Haptics {
    static func tap() { UIImpactFeedbackGenerator(style: .light).impactOccurred() }
    static func select() { UISelectionFeedbackGenerator().selectionChanged() }
    static func firm() { UIImpactFeedbackGenerator(style: .medium).impactOccurred() }
    static func error() { UINotificationFeedbackGenerator().notificationOccurred(.error) }
}

File: android/app/src/main/java/cz/drabek/chuvicka/ui/Theme.kt
package cz.drabek.chuvicka.ui

import androidx.compose.foundation.Canvas
import androidx.compose.foundation.isSystemInDarkTheme
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.darkColorScheme
import androidx.compose.material3.lightColorScheme
import androidx.compose.runtime.Composable
import androidx.compose.runtime.CompositionLocalProvider
import androidx.compose.runtime.staticCompositionLocalOf
import androidx.compose.ui.Modifier
import androidx.compose.ui.geometry.CornerRadius
import androidx.compose.ui.geometry.Offset
import androidx.compose.ui.geometry.Size
import androidx.compose.ui.graphics.Color
import cz.drabek.chuvicka.Settings
import cz.drabek.chuvicka.parent.RoomState
import kotlin.math.max

/** The palette of the iOS app: calm night blues, a moon yellow, and warm colours for sound. */
data class Palette(
    val sky: Color, val card: Color, val ink: Color, val muted: Color,
    val moon: Color, val accent: Color, val calm: Color, val warn: Color,
    val alarm: Color, val loud: Color, val middle: Color, val neutral: Color,
    /** The dark appearance: the fields use the dark tokens, with white type on all. */
    val dark: Boolean = false,
)

val LightPalette = Palette(
    sky = Color(0xFFF6F8FC), card = Color(0xFFFFFFFF), ink = Color(0xFF141A2B), muted = Color(0xFF5E6679),
    moon = Color(0xFFF5CF6E), accent = Color(0xFFA86B0C), calm = Color(0xFF1C9A6C), warn = Color(0xFFD27D12),
    alarm = Color(0xFFD8433D), loud = Color(0xFFE0613F), middle = Color(0xFFE3A020), neutral = Color(0xFF9AA3B8),
)

val DarkPalette = Palette(
    sky = Color(0xFF090E1F), card = Color(0xFF151B33), ink = Color(0xFFEEF1F8), muted = Color(0xFF9AA3B8),
    moon = Color(0xFFF7D98C), accent = Color(0xFFF7D98C), calm = Color(0xFF73E3B8), warn = Color(0xFFFFB861),
    alarm = Color(0xFFFF6B6B), loud = Color(0xFFFF9973), middle = Color(0xFFF7D98C), neutral = Color(0xFF5A6070),
    dark = true,
)

/**
 * The colours of the room state: the field behind the glance view, the band and the frame of the
 * picture view, the waveform bars. The same values as on the iPhone. They never change with the
 * wallpaper: no dynamic colour, so a purple wallpaper does not give a purple „Klid".
 */
object StateColors {
    // Light fields: a lit room, by day.
    val calm = Color(0xFF00A1A0)          // teal, white type
    val sound = Color(0xFFFBC040)         // amber, ink type
    val cry = Color(0xFFA81233)           // wine, white type
    val graphite = Color(0xFF3A3D45)      // lost and connecting, white type
    // Dark fields: the dark appearance, and the light one after 30 s untouched. White type on all.
    val calmDim = Color(0xFF117376)
    val soundDim = Color(0xFF78662E)
    val cryDim = Color(0xFF570F29)
    val graphiteDim = Color(0xFF2A2C33)
    /** The type on amber. White there would be about 2:1. */
    val ink = Color(0xFF1B1B1F)
    /** The ribbon (muted, volume low, a short gap): never a field colour. White type, 5.4:1. */
    val ribbon = Color(0xFF5C6B8A)
    /** The glyph of „Nehlídá" on graphite: the dark-mode alarm red, 4.0:1. */
    val lostGlyph = Color(0xFFFF6B6B)
    /** On black, in Night mode: the state colour as a dim accent. Wine is too dark on black. */
    val cryNight = Color(0xFFC64B70)
}

/** The field of a state. [dim]: after 30 s untouched. The dark appearance is always dim. */
fun Palette.field(state: RoomState, dim: Boolean = false): Color {
    val d = dim || dark
    return when (state) {
        RoomState.CALM -> if (d) StateColors.calmDim else StateColors.calm
        RoomState.SOUND -> if (d) StateColors.soundDim else StateColors.sound
        RoomState.CRY -> if (d) StateColors.cryDim else StateColors.cry
        RoomState.LOST, RoomState.CONNECTING -> if (d) StateColors.graphiteDim else StateColors.graphite
    }
}

/** The glyph and the word on a field: white, only ink on the bright amber. */
fun Palette.onField(state: RoomState, dim: Boolean = false): Color =
    if (state == RoomState.SOUND && !dim && !dark) StateColors.ink else Color.White

/** Small text on a field: white 90 %, ink 70 % on amber. Aims at 4.5:1. */
fun Palette.onFieldSecondary(state: RoomState, dim: Boolean = false): Color {
    val on = onField(state, dim)
    return on.copy(alpha = if (on == Color.White) 0.9f else 0.7f)
}

/** The state colour on black (Night mode), before the alpha of the glyph or the word. */
fun nightAccent(state: RoomState): Color = when (state) {
    RoomState.CALM -> StateColors.calm
    RoomState.SOUND -> StateColors.sound
    RoomState.CRY -> StateColors.cryNight
    RoomState.LOST -> StateColors.lostGlyph
    RoomState.CONNECTING -> Color.White
}

val LocalPalette = staticCompositionLocalOf { LightPalette }

val colors: Palette @Composable get() = LocalPalette.current

@Composable
fun ChuvickaTheme(appearance: Settings.Appearance, forceDark: Boolean = false, content: @Composable () -> Unit) {
    val dark = forceDark || when (appearance) {
        Settings.Appearance.LIGHT -> false
        Settings.Appearance.DARK -> true
        Settings.Appearance.AUTO -> isSystemInDarkTheme()
    }
    val p = if (dark) DarkPalette else LightPalette
    val scheme = if (dark) darkColorScheme(
        primary = p.moon, onPrimary = Color.Black, background = p.sky, surface = p.card,
        onBackground = p.ink, onSurface = p.ink, error = p.alarm, secondary = p.accent,
    ) else lightColorScheme(
        primary = p.accent, onPrimary = Color.White, background = p.sky, surface = p.card,
        onBackground = p.ink, onSurface = p.ink, error = p.alarm, secondary = p.accent,
    )
    CompositionLocalProvider(LocalPalette provides p) {
        MaterialTheme(colorScheme = scheme, content = content)
    }
}

/** The colour of one bar: the colours of the states, so the waveform and the field agree. */
fun Palette.level(v: Float): Color = when {
    v < 0.35f -> StateColors.calm
    v < 0.7f -> if (dark) StateColors.sound else Color(0xFFE3A020)    // A deeper amber reads on the light sky.
    else -> if (dark) StateColors.cryNight else StateColors.cry
}

/**
 * The last 6 seconds of sound, as bars from the middle. [tint]: one colour for all bars, for a
 * waveform on a field (teal bars on a teal field would not show); the height carries the level.
 */
@Composable
fun Waveform(history: List<Float>, modifier: Modifier = Modifier, dim: Boolean = false, tint: Color? = null) {
    val p = colors
    Canvas(modifier.fillMaxWidth()) {
        val n = history.size
        val gap = size.width / n
        val bar = max(2f, gap * 0.45f)
        for ((i, v) in history.withIndex()) {
            val h = max(bar, v * size.height)
            val alpha = if (dim) 0.35f else 0.35f + 0.65f * (i.toFloat() / n)
            drawRoundRect(
                color = (tint ?: if (dim) p.neutral else p.level(v)).copy(alpha = alpha),
                topLeft = Offset(i * gap + (gap - bar) / 2, (size.height - h) / 2),
                size = Size(bar, h),
                cornerRadius = CornerRadius(bar / 2, bar / 2),
            )
        }
    }
}

/** "před 7 s", "před 3 min": the Czech relative form. */
fun ago(millis: Long, now: Long = System.currentTimeMillis()): String {
    val s = max(0L, (now - millis) / 1000)
    return when {
        s < 60 -> "před $s s"
        s < 3600 -> "před ${s / 60} min"
        else -> "před ${s / 3600} h"
    }
}
